File: python/bowling_game.py
class BowlingGame:
    def __init__(self) -> None:
        self.rolls = []

    def roll(self, pins):
        self.rolls.append(pins)

    def score(self):
        result = 0
        rollIndex = 0
        for frameIndex in range(10):
            if self._isStrike(rollIndex):
                result += self._strikeScore(rollIndex)
                rollIndex += 1
            elif self._isSpare(rollIndex):
                result += self._spareScore(rollIndex)
                rollIndex += 2
            else:
                result += self._frameScore(rollIndex)
                rollIndex += 2
        return result

    def _isStrike(self, rollIndex):
        return self.rolls[rollIndex] == 10

    def _isSpare(self, rollIndex):
        return self.rolls[rollIndex] + self.rolls[rollIndex+1] == 10

    def _strikeScore(self, rollIndex):
        return 10 + self.rolls[rollIndex+1] + self.rolls[rollIndex+2]

    def _spareScore(self, rollIndex):
        return 10 + self.rolls[rollIndex+2]

    def _frameScore(self, rollIndex):
        return self.rolls[rollIndex] + self.rolls[rollIndex+1]

File: cpp/bowling_game/bowling_game.cc
#include "bowling_game.h"

void BowlingGame::Roll(int pins) { rolls_[current_roll_++] = pins; }

int BowlingGame::Score()
{
    int score = 0;
    int frame_index = 0;
    for (int frame = 0; frame < 10; frame++)
    {
        if (IsStrike(frame_index))
        {
            score += 10 + StrikeBonus(frame_index);
            frame_index++;
        }
        else if (IsSpare(frame_index))
        {
            score += 10 + SpareBonus(frame_index);
            frame_index += 2;
        }
        else
        {
            score += SumOfBallsInFrame(frame_index);
            frame_index += 2;
        }
    }
    return score;
}

bool BowlingGame::IsStrike(int frame_index) { return rolls_[frame_index] == 10; }

bool BowlingGame::IsSpare(int frame_index)
{
    return rolls_[frame_index] + rolls_[frame_index + 1] == 10;
}

int BowlingGame::SumOfBallsInFrame(int frame_index)
{
    return rolls_[frame_index] + rolls_[frame_index + 1];
}

int BowlingGame::SpareBonus(int frame_index) { return rolls_[frame_index + 2]; }

int BowlingGame::StrikeBonus(int frame_index)
{
    return rolls_[frame_index + 1] + rolls_[frame_index + 2];
}